### ISSUE TITLE
Delayed_job gets stuck and does not process jobs

### ISSUE THREAD

**Issue by sauloperez** [openfoodfoundation/ofn-install#119]
## What? Why?
A customer complained about the shop page being empty although the order cycle and the products where correctly populated. We then found out it was due to cache not being refreshed.

We then manually triggered a cache refresh executing `OpenFoodNetwork::ProductsCacheRefreshment.refresh(enterprise, order_cycle)` from the Rails console, but the job kept enqueued but not processed and the log had the following contents:

```
openfoodnetwork@scw-ecce22:~/apps/openfoodnetwork/current$ tail -f log/delayed_job.log                                 
2018-02-10T04:59:48+0000: [Worker(delayed_job host:scw-ecce22 pid:4035)] Error while reserving job: PG::Error: connection not open    
: UPDATE "delayed_jobs" SET locked_at = '2018-02-10 04:59:48.869453', locked_by = 'delayed_job host:scw-ecce22 pid:4035' WHERE id IN (SELECT
  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-02-10 04:59:48.866440' AND (locked_at IS NULL OR locked_at < '2018-02-10 04:44:48.866533')
 OR locked_by = 'delayed_job host:scw-ecce22 pid:4035') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNI
NG *                                                                                                                                  
2018-02-10T04:59:53+0000: [Worker(delayed_job host:scw-ecce22 pid:4035)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-02-10 04:59:53.876770', locked_by = 'delayed_job host:scw-ecce22 pid:4035' WHERE id IN (SELECT
  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-02-10 04:59:53.873866' AND (locked_at IS NULL OR locked_at < '2018-02-10 04:44:53.873958')
 OR locked_by = 'delayed_job host:scw-ecce22 pid:4035') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNI
NG *
2018-02-10T04:59:58+0000: [Worker(delayed_job host:scw-ecce22 pid:4035)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-02-10 04:59:58.883130', locked_by = 'delayed_job host:scw-ecce22 pid:4035' WHERE id IN (SELECT
  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-02-10 04:59:58.880787' AND (locked_at IS NULL OR locked_at < '2018-02-10 04:44:58.880871')
 OR locked_by = 'delayed_job host:scw-ecce22 pid:4035') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNI
NG *
2018-02-10T05:00:03+0000: [Worker(delayed_job host:scw-ecce22 pid:4035)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-02-10 05:00:03.890062', locked_by = 'delayed_job host:scw-ecce22 pid:4035' WHERE id IN (SELECT
  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-02-10 05:00:03.887335' AND (locked_at IS NULL OR locked_at < '2018-02-10 04:45:03.887422')
 OR locked_by = 'delayed_job host:scw-ecce22 pid:4035') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNI
NG *
2018-02-10T05:00:08+0000: [Worker(delayed_job host:scw-ecce22 pid:4035)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-02-10 05:00:08.897064', locked_by = 'delayed_job host:scw-ecce22 pid:4035' WHERE id IN (SELECT
  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-02-10 05:00:08.894194' AND (locked_at IS NULL OR locked_at < '2018-02-10 04:45:08.894283')
 OR locked_by = 'delayed_job host:scw-ecce22 pid:4035') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNI
NG *
```

Once we started it with `RAILS_ENV=production script/delayed_job restart` the issue got fixed, processing the jobs as follows:

```
openfoodnetwork@scw-ecce22:~/apps/openfoodnetwork/current$ tail -f log/delayed_job.log
2018-02-11T11:29:23+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1514) FAILED (18 prior attempts) with ActiveRecord::RecordNotFound: Couldn't find OrderCycle with id=92
2018-02-11T11:29:23+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1557) RUNNING
2018-02-11T11:29:23+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1557) COMPLETED after 0.5753
2018-02-11T11:29:23+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1558) RUNNING
2018-02-11T11:29:24+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1558) COMPLETED after 0.7385
2018-02-11T11:29:24+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1559) RUNNING
2018-02-11T11:29:25+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1559) COMPLETED after 0.5659
2018-02-11T11:29:25+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1561) RUNNING
2018-02-11T11:29:25+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] Job RefreshProductsCacheJob (id=1561) COMPLETED after 0.1507
2018-02-11T11:29:25+0000: [Worker(delayed_job host:scw-ecce22 pid:31288)] 7 jobs processed at 0.3704 j/s, 1 failed
```

## What next?

So far, delayed job is the most unreliable part of the infrastructure leading to these hard to spot issues. I wonder whether we should restart it after each deployment (if we don't do it already) but also dig deeper on these particular issues to find out what the actual cause might be.

EDIT Yes, we do but trusting Monit to start it for us.
https://github.com/openfoodfoundation/ofn-install/blob/19f5e8108e70f0395176387215af8fc9b58ff65a/roles/deploy/handlers/main.yml#L16

**Comment by sauloperez** [comment]
According to the customer, the hub manager duplicates the previous order cycle, then the producer selects the product he has available, and then the hub manager adds them to the order cycle.

If the cache does not refresh in the last step, we have right there a shop fetching an empty products list from the cache, although they will be in the DB.

**Comment by Matt-Yorkley** [comment]
I think we do restart it, but I've found a few references to this sort of issue whilst digging for answers the other day. It looks like when the service is restarted there can sometimes be old delayed_job processes left running, which can cause issues: 
https://stackoverflow.com/questions/1971315/delayed-job-performs-not-up-to-date-code/1975155#1975155 

Or sometimes the delayed job is working with instances of stale objects that don't correspond to codebase changes:
https://stackoverflow.com/questions/4903375/delayed-job-how-to-reload-the-payload-classes-during-every-call-in-development

**Comment by pacodelaluna** [comment]
I don't know why but I have an error on this recipe when I am deploying ... So I have to restart delayed_job manually ...
```
fatal: [staging.openfoodfrance.fr]: FAILED! => {
    "changed": true, 
    "cmd": [
        "bash", 
        "-lc", 
        "./script/delayed_job -i 0 stop RAILS_ENV=staging"
    ], 
    "delta": "0:00:07.338070", 
    "end": "2018-03-07 15:52:30.640147", 
    "failed": true, 
    "invocation": {
        "module_args": {
            "_raw_params": "bash -lc \"./script/delayed_job -i 0 stop RAILS_ENV=staging\"", 
            "_uses_shell": false, 
            "chdir": "/home/offrance/apps/openfoodnetwork/current", 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }, 
        "module_name": "command"
    }, 
    "rc": 1, 
    "start": "2018-03-07 15:52:23.302077", 
    "stderr": "[SKYLIGHT] [1.5.0] Running Skylight in development mode. No data will be reported until you deploy your app.\n(To disable this message for all local apps, run `skylight disable_dev_warning`.)\n/home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:1222:in `initialize': fe_sendauth: no password supplied (PG::Error)\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:1222:in `new'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:1222:in `connect'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:324:in `initialize'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:28:in `new'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:28:in `postgresql_connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:315:in `new_connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:325:in `checkout_new_connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:247:in `block (2 levels) in checkout'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:242:in `loop'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:242:in `block in checkout'\n\tfrom /home/offrance/.rbenv/versions/2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:239:in `checkout'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:102:in `block in connection'\n\tfrom /home/offrance/.rbenv/versions/2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:101:in `connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:410:in `retrieve_connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_specification.rb:171:in `retrieve_connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_specification.rb:145:in `connection'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/activerecord-3.2.21/lib/active_record/model_schema.rb:224:in `table_exists?'\n\tfrom /home/offrance/.gem/ruby/2.1.0/bundler/gems/spree-86bf87f1b1e1/core/app/models/spree/preferences/store.rb:89:in `should_persist?'\n\tfrom /home/offrance/.gem/ruby/2.1.0/bundler/gems/spree-86bf87f1b1e1/core/app/models/spree/preferences/store.rb:73:in `persist'\n\tfrom /home/offrance/.gem/ruby/2.1.0/bundler/gems/spree-86bf87f1b1e1/core/app/models/spree/preferences/store.rb:21:in `set'\n\tfrom /home/offrance/.gem/ruby/2.1.0/bundler/gems/spree-86bf87f1b1e1/core/app/models/spree/preferences/preferable_class_methods.rb:27:in `block in preference'\n\tfrom /home/offrance/.gem/ruby/2.1.0/bundler/gems/spree-86bf87f1b1e1/core/app/models/spree/preferences/preferable.rb:41:in `set_preference'\n\tfrom /home/offrance/apps/openfoodnetwork/current/config/application.rb:38:in `block in <class:Application>'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:30:in `instance_exec'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:30:in `run'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:55:in `block in run_initializers'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:54:in `each'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:54:in `run_initializers'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/application.rb:136:in `initialize!'\n\tfrom /home/offrance/.gem/ruby/2.1.0/gems/railties-3.2.21/lib/rails/railtie/configurable.rb:30:in `method_missing'\n\tfrom /home/offrance/apps/openfoodnetwork/current/config/environment.rb:5:in `<top (required)>'\n\tfrom ./script/delayed_job:3:in `require_relative'\n\tfrom ./script/delayed_job:3:in `<main>'", 
    "stdout": "", 
    "stdout_lines": [], 
    "warnings": []
}
```

**Comment by sauloperez** [comment]
I got another case and by checking the log it seems to the same case as in the description:

```
2018-03-07T10:39:21+0000: [Worker(delayed_job host:scw-ecce22 pid:18937)] Error while reserving job: PG::Error: server closed the connection unexpectedly                                                                                                                                       This probably means the server terminated abnormally
        before or while processing the request.                                                                                    
: UPDATE "delayed_jobs" SET locked_at = '2018-03-07 10:39:20.999864', locked_by = 'delayed_job host:scw-ecce22 pid:18937' WHERE id IN (SELECT  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-03-07 10:39:20.996961' AND (locked_at IS NULL OR locked_at < '2018-03-07 10:24:20.997043') OR locked_by = 'delayed_job host:scw-ecce22 pid:18937') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNING *
2018-03-07T10:39:26+0000: [Worker(delayed_job host:scw-ecce22 pid:18937)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-03-07 10:39:26.132597', locked_by = 'delayed_job host:scw-ecce22 pid:18937' WHERE id IN (SELECT  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-03-07 10:39:26.130077' AND (locked_at IS NULL OR locked_at < '2018-03-07 10:24:26.130177') OR locked_by = 'delayed_job host:scw-ecce22 pid:18937') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNING *
2018-03-07T10:39:31+0000: [Worker(delayed_job host:scw-ecce22 pid:18937)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-03-07 10:39:31.138596', locked_by = 'delayed_job host:scw-ecce22 pid:18937' WHERE id IN (SELEC
T  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-03-07 10:39:31.136135' AND (locked_at IS NULL OR locked_at < '2018-03-07 10:24:31.136218'
) OR locked_by = 'delayed_job host:scw-ecce22 pid:18937') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETUR
NING *
2018-03-07T10:39:36+0000: [Worker(delayed_job host:scw-ecce22 pid:18937)] Error while reserving job: PG::Error: connection not open
: UPDATE "delayed_jobs" SET locked_at = '2018-03-07 10:39:36.145681', locked_by = 'delayed_job host:scw-ecce22 pid:18937' WHERE id IN (SELEC
T  id FROM "delayed_jobs"  WHERE ((run_at <= '2018-03-07 10:39:36.142417' AND (locked_at IS NULL OR locked_at < '2018-03-07 10:24:36.142500'
) OR locked_by = 'delayed_job host:scw-ecce22 pid:18937') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETUR
NING *
```

**Comment by enricostano** [comment]
```
==> /home/openfoodnetwork/apps/openfoodnetwork/current/log/delayed_job.log <==
2018-03-13T17:11:57+0100: [Worker(delayed_job host:ofn-de pid:21631)] Job Spree::User#send_on_create_confirmation_instructions_without_delay (id=5) RUNNING

==> /home/openfoodnetwork/apps/openfoodnetwork/current/log/production.log <==
127.0.0.1:11211 failed (count: 0) RuntimeError: Cannot share client between multiple processes
  Rendered shared/mailers/_signoff.html.haml (1.9ms)
  Rendered shared/mailers/_social_and_contact.html.haml (7.4ms)
  Rendered spree/user_mailer/confirmation_instructions.html.haml within layouts/mailer (18.3ms)
  Rendered spree/user_mailer/confirmation_instructions.text.erb (1.9ms)
```

It seems that we have some problem connecting to Memcached while sending emails :thinking:

**Comment by Matt-Yorkley** [comment]
Closed by #192
